### PR TITLE
[WIP] Block Supports: Try stabilizing default controls to top-level property

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/pull/67018
- https://github.com/WordPress/gutenberg/pull/45994

⚠️ ⚠️ ⚠️**This PR is for discussion purposes only and isn't intended to be merged** ⚠️ ⚠️⚠️ 

## What?

This PR is a proof of concept to see what it would take to stabilize current `supports.feature.__experimentalDefaultControls` config to its own top-level `defaultControls` property instead of being stabilized within the `supports.feature` object.

## Why?

There's a desire to configure non-block-support controls via the `defaultControls` property. This makes a touch more sense if it is a top-level block type property.

## How?

**Requires core updates to works (at least to my current knowledge)**

- When stabilizing `__experimentalDefaults` move the config to a new top-level `defaultControls` property
- If existing config for a feature within `defaultControls` exists. That already stable config in `defaultControls` will be used
- Unlike the current approach to stabilization within `supports`, this approach can't determine the order the experimental or stable default controls config was defined in so is slightly more "dumb" and would be a little more prone to ignoring plugin or theme filtered default controls. It is also then an exception to how the rest of the experimental block supports are stabilized.
- This update will require extenders to do more than just switch keys from `__experimentalDefaultControls` to `defaultControls` within a current `getBlockSupport()` call. Now, the config no longer exists within block supports, they'll need to update to calling a new util `getBlockDefaultControls`.

## Testing Instructions

Proper testing instructions are TBA once it's confirmed we want to go in this direction.

1. Apply core patch from diff below
<details>
<summary>Core diff needed (I think 🙂 )</summary>

```diff
diff --git a/src/wp-admin/includes/post.php b/src/wp-admin/includes/post.php
index e5e1e5d147..0a16e08bd2 100644
--- a/src/wp-admin/includes/post.php
+++ b/src/wp-admin/includes/post.php
@@ -2317,6 +2317,7 @@ function get_block_editor_server_block_settings() {
 		'example'          => 'example',
 		'variations'       => 'variations',
 		'allowed_blocks'   => 'allowedBlocks',
+		'default_controls' => 'defaultControls',
 	);
 
 	foreach ( $block_registry->get_all_registered() as $block_name => $block_type ) {
diff --git a/src/wp-includes/blocks.php b/src/wp-includes/blocks.php
index 94c4dfa43a..7b954aa971 100644
--- a/src/wp-includes/blocks.php
+++ b/src/wp-includes/blocks.php
@@ -486,6 +486,7 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 		'variations'      => 'variations',
 		'example'         => 'example',
 		'allowedBlocks'   => 'allowed_blocks',
+		'defaultControls' => 'default_controls',
 	);
 	$textdomain        = ! empty( $metadata['textdomain'] ) ? $metadata['textdomain'] : null;
 	$i18n_schema       = get_block_metadata_i18n_schema();
diff --git a/src/wp-includes/class-wp-block-type.php b/src/wp-includes/class-wp-block-type.php
index 6916f6a462..5941fc8c53 100644
--- a/src/wp-includes/class-wp-block-type.php
+++ b/src/wp-includes/class-wp-block-type.php
@@ -150,6 +150,14 @@ class WP_Block_Type {
 	 */
 	public $supports = null;
 
+	/**
+	 * Default UI controls.
+	 *
+	 * @since 6.8.0
+	 * @var array|null
+	 */
+	public $default_controls = null;
+
 	/**
 	 * Structured data for the block preview.
 	 *
diff --git a/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php b/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
index 263f2a8ef2..947003874e 100644
--- a/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
+++ b/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
@@ -298,6 +298,7 @@ class WP_REST_Block_Types_Controller extends WP_REST_Controller {
 				'view_style_handles',
 				'variations',
 				'block_hooks',
+				'default_controls',
 			),
 			$deprecated_fields
 		);
@@ -762,6 +763,14 @@ class WP_REST_Block_Types_Controller extends WP_REST_Controller {
 					'context'           => array( 'embed', 'view', 'edit' ),
 					'readonly'          => true,
 				),
+				'default_controls'       => array(
+					'description' => __( 'UI controls to display by default.' ),
+					'type'        => 'object',
+					'default'     => array(),
+					'properties'  => array(),
+					'context'     => array( 'embed', 'view', 'edit' ),
+					'readonly'    => true,
+				),
 			),
 		);
 

```
</details>

2. Checkout this PR
3. Edit a post and add in several blocks that contain different default control configurations
4. Crosscheck the controls shown by default in the block inspector against the block's block.json configuration.

## Screenshots or screencast 

There should be no visual difference if this PR hasn't broken something 😅

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
@youknowriad this PoC should give some indication as to what is required to move default controls configuration to a top-level property. 

I got a little lost trying to get something working here, so there's a good chance I'm missing something, or there's a better way of doing this. That uncertainty and going in circles trying to understand how our block types are handled across the server and editor doesn't leave me with much confidence to foresee issues with this approach.

The biggest catches I've found so far are:

1. Unlike the other experimental block supports that have been stabilized, with this approach to default controls we have even less ability to determine the correct value after filters have been applied to block type args.
2. Due to the above, the stabilization of default controls would behave slightly differently from the other recently stabilised experimental supports.
3. If we can't find an alternative approach to the core updates I found I needed, to get `defaultControls` as a top-level property working, would we need to delay stabilizing `defaultControls` until those core updates are in the minimum supported version of WP?

As noted earlier, I'm a bit confused around what exactly is needed on the core side of things right now, so take that last point with a grain of salt.

I can take a closer look at this tomorrow with a fresh set of eyes and mind. Following that though, I'll be AFK on and off for the next couple of weeks. It might take me some time to wrangle this if we decide to go ahead with the top-level property.

Any thoughts?

**Comment by youknowriad** [comment]
It seems some kind of hacky backwards compatibility code is needed regardless of where the feature end up. I would love a recap of the "public" APIs that we need to maintain backward compatibility for. It's a bit hard to read the changes here, but my understanding is that we're shiming in multiple places and ideally we should only do it in a single place.

**Comment by aaronrobertshaw** [comment]
> It seems some kind of hacky backwards compatibility code is needed regardless of where the feature end up

It felt that way during my explorations today. 

There are several things required to introduce a top-level block type property on the core side (see diff in PR description) as well as in Gutenberg (see https://github.com/WordPress/gutenberg/pull/67729/commits/06a44a96303ec7b7823ba0737099d009bde78678). 

On the core side I needed to add `defaultControls`/`default_controls` to the:
- [fields to pick](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/post.php?plain=1#L2319) in the block editor server block settings
- [property mappings](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks.php?plain=1#L488) within `register_block_type_from_metadata`
- [`WP_Block_Type` class](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-block-type.php?plain=1#L145-L151)
- [extra fields](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php?plain=1#L273) and [schema](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php?plain=1#L469) for the block types REST API controller

I don't know how easy it would be, or even if it's possible, to override all that within Gutenberg for backwards compatibility.

>  I would love a recap of the "public" APIs that we need to maintain backward compatibility for.

I'll give it a go, if I've gone down the wrong track here let me know 🙂 

The current plan for stabilization of experimental supports started with https://github.com/WordPress/gutenberg/pull/63401, which involved stabilizing the block supports :

- on the PHP side via a [filter](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.8/blocks.php?plain=1#L18) in Gutenberg or the [`WP_Block_Type` class in core](https://github.com/WordPress/wordpress-develop/pull/7069)
- the JS side via the [`processBlockType`](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/store/process-block-type.js?plain=1#L230) action


Experimental support keys were to be replaced with stable counterparts. The following key maps reflect what is being stabilized in these overall efforts.

```js
export const EXPERIMENTAL_SUPPORTS_MAP = {
	__experimentalBorder: 'border',
};

export const COMMON_EXPERIMENTAL_PROPERTIES = {
	__experimentalDefaultControls: 'defaultControls',
	__experimentalSkipSerialization: 'skipSerialization',
};

export const EXPERIMENTAL_SUPPORT_PROPERTIES = {
	typography: {
		__experimentalFontFamily: 'fontFamily',
		__experimentalFontStyle: 'fontStyle',
		__experimentalFontWeight: 'fontWeight',
		__experimentalLetterSpacing: 'letterSpacing',
		__experimentalTextDecoration: 'textDecoration',
		__experimentalTextTransform: 'textTransform',
	},
};
```

The bulk of those should be relatively painless because it is only the key changing, not where they live. For extenders developing custom blocks that opt into these block supports they won't need to change anything immediately. Ideally, though, they'll migrate their custom block's block.json or block type definitions to use the stable support keys.

If 3rd parties are filtering block type args or metadata, they'll continue to function for the most part. The main edge case here is where there are multiple plugins/themes filtering the same supports using a mix of experimental and stable keys to change the same support. That scenario can't be accounted for 100% so the approach taken will check the insertion order in the config to make a "best guess" as to which value was updated last and use that.

If plugins or themes are checking block supports in custom code, they'll need to update that code to check the stable key.

This all would apply to `defaultControls` too if it remains within `supports`. Moving it to a sibling of `supports` instead means not only having to wrangle adding the new block type property and making sure that works in Gutenberg for older WordPress releases, but also requiring some extra changes to minimise the updates forced upon plugins and themes e.g. how they check for default controls.

During my tinkering today, I also looked at monkey patching `getBlockSupport` to retrieve the stabilized support values but shelved it as too great a hack. It could be revisited, or an alternative found if we want to limit the need for plugins to update keys in their custom checks for block supports.

FWIW a [quick search](https://wpdirectory.net/search/01JDPSPQ4NG60QMAXKPP9WPVWA) suggests that 1656 plugins are using `__experimentalDefaultControls` in some shape or form.

I hope that wall of text helps paint a clearer picture 🙏

**Comment by youknowriad** [comment]
> I don't know how easy it would be, or even if it's possible, to override all that within Gutenberg for backwards compatibility.

I think these changes are ok (the core ones) but I do believe we should be able to do that in Gutenberg, unrelated to defaultControls. it shouldn't be a pain point for us to introduce a new optional key to block.json

> If 3rd parties are filtering block type args or metadata, they'll continue to function for the most part. The main edge case here is where there are multiple plugins/themes filtering the same supports using a mix of experimental and stable keys to change the same support. That scenario can't be accounted for 100% so the approach taken will check the insertion order in the config to make a "best guess" as to which value was updated last and use that.

My main thinking here is that we should **not** be patching the "defaultControls" at registration time (both server and client). (I guess the same reasoning can apply to block supports and any deprecated property). Instead, we should be patching the selectors we use to retrieve these. In simpler words:

 - The selectors (both php and JS) to retrieve the default controls should look in both places (stable and deprecated ones) and "merge them". If deprecated places are found, we should trigger a deprecated call in the selector.

Doing this at the selector level prevents any potential hacks we need to do to support registration/filtering... 

--- 
These are obviously just theoretical thoughts, maybe there's a flaw in this reasoning as I know these things are complex. Also it seems the same issues apply to "defaultControls" and "block supports" stabilization regardless of whether the stable keys live, so I would say this cumbersome deprecation shouldn't be impacting the decision on where to put "defaultControls" (Correct me if I'm wrong).

It seems the main difference between top-level or not, is the difficulty to add a top level property in Gutenberg (overriding Core).

**Comment by aaronrobertshaw** [comment]
Appreciate the continued guidance @youknowriad, thanks 🙇 

>  we should be able to do that in Gutenberg, unrelated to defaultControls. it shouldn't be a pain point for us to introduce a new optional key to block.json

Agreed, it shouldn't be a pain point. I'd say a large part of the issue is I'm not very familiar with this area so it's tougher to know what needs doing and how. I'm hopeful I can figure it out, it might just take some time.

> Doing this at the selector level prevents any potential hacks we need to do to support registration/filtering...

If I understand correctly, we'll face the same edge cases around filtered values. Wherever we fetch and merge the stable and experimental values, we need to decide on how that is done. 

- Do we continue with the current approach of using the insertion order of experimental vs stable keys to mitigate potentially losing experimental key values set by filters, while plugins migrate to stable keys?
- Should we have opted for a simpler, clearer, but more limited option of preferring stable key values over experimental ones? 
 
> I would say this cumbersome deprecation shouldn't be impacting the decision on where to put "defaultControls" (Correct me if I'm wrong).

If the decision purely comes down to where makes the most sense for `defaultControls`. The expansion of that API's scope to include non-block-support controls tilts the decision in favour of the top-level property, in my opinion. 

The hope with the block supports stabilization was to have this all in place early in the 6.8 release cycle. If we revisit the approach to handle all this within selectors (which I'm on board with), it will push back when this all lands until into the new year.

> It seems the main difference between top-level or not, is the difficulty to add a top level property in Gutenberg (overriding Core).

That's it in a nutshell.

**Comment by aaronrobertshaw** [comment]
> It seems the main difference between top-level or not, is the difficulty to add a top level property in Gutenberg (overriding Core).

I've taken another run at overriding core to add the default controls property within Gutenberg. In https://github.com/WordPress/gutenberg/pull/67729/commits/0f4914acd5e1d9d6c5db6670fbe81349688a94d8 I've managed to avoid all but one core update. 

What I haven't been able to solve yet is updating the [fields to pick](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/post.php?plain=1#L2300-L2320) in core's `get_block_editor_server_block_settings` function. It seems the results of that function are primarily used to provide JSON encoded block type data to `wp.blocks.unstable__bootstrapServerSideBlockDefinitions()` calls, e.g. [here for the block editor](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/edit-form-blocks.php?plain=1#L107-L111).

After https://github.com/WordPress/gutenberg/pull/67729/commits/0f4914acd5e1d9d6c5db6670fbe81349688a94d8 and very quick initial tests, it appears that if I define `defaultControls` in a Gutenberg block's block.json file, that data is available and used correctly. 

Given that, my latest plan is to remove the stabilization of experimental block supports from the `register_block_type_args` filter and JS `processBlockType` and instead add deprecation warnings in those places if a block type contains a deprecated experimental support flag. Following that, I'll explore the proposed approach and update the JS selectors and PHP util functions to collect and merge the experimental and stable values, trigger deprecation warnings etc.

Happy for nudges in the right direction if I'm heading down the wrong track.

**Comment by ramonjd** [comment]
Thanks for looking into this @aaronrobertshaw and exploring the options.

> The hope with the block supports stabilization was to have this all in place early in the 6.8 release cycle. If we revisit the approach to handle all this within selectors (which I'm on board with), it will push back when this all lands until into the new year.

That sounds like a good option if it takes the pressure off, and gives folks time to come up with robust approach.

We haven't committed the [core patch yet](https://github.com/WordPress/wordpress-develop/pull/7069), so it can be put on ice at any time.

**Comment by youknowriad** [comment]
Your proposed plan sounds good to me.

> Do we continue with the current approach of using the insertion order of experimental vs stable keys to mitigate potentially losing experimental key values set by filters, while plugins migrate to stable keys?
> Should we have opted for a simpler, clearer, but more limited option of preferring stable key values over experimental ones?

Insertion order seems as arbitrary as preferring stable keys no? What are we hoping to gain by doing this?

**Comment by aaronrobertshaw** [comment]
> Insertion order seems as arbitrary as preferring stable keys no? What are we hoping to gain by doing this?

It is arbitrary, true, but a little more flexible than simply preferring the stable key values.

The plan to date has been to update all the core blocks to use the stable block supports as soon as this stabilization is solid. That is partly due to our block library effectively forming part of our overall documentation and setting examples for 3rd parties. 

If we prefer stable over experimental values, all plugins that filter these supports would see their filters fail to take effect. Using the insertion order helps alleviate that for the majority of use cases. 

It doesn't solve an edge case where multiple filters of the same support config use a mix of experimental and stable keys, however, this scenario is much more contrived than a plugin being slow to migrate its filters to use stable keys.

In other words, by using the insertion order, we can provide some backward compatibility for the following situation:

1. Core blocks use stable block support keys
2. Plugin or theme filters a core block to disable a particular block support using the experimental support key
3. The expected filtered value is still used

Given that scenario, do you have a preferred approach?

**Comment by youknowriad** [comment]
> If we prefer stable over experimental values, all plugins that filter these supports would see their filters fail to take effect. Using the insertion order helps alleviate that for the majority of use cases.

What if the order is reversed?

**Comment by aaronrobertshaw** [comment]
> What if the order is reversed?

Do you mean our order of preference? So preferring experimental key values over stable?

We'd have a similar scenario/problem.

1. Custom block uses experimental key
2. Theme filters the block supports using the (soon to be) documented stable key
3. The filtered value that used the stable key wouldn't take effect

Using the insertion order approach with the above scenario:

1. The block would start with the experimental key for its support
2. After the filter is applied it will now have both the experimental and stable support keys in that order.
3. Preferring the last inserted key as our best guess for the last updated value, the filter's stable key value would be used matching the expected outcome

Likewise, using the other scenario:

- Core blocks use stable block support keys
- Plugin or theme filters a core block to disable a particular block support using the experimental support key
- The expected filtered value is still used

The block starts with the stable key, then gets the experimental afterwards via the filter. The experimental key is inserted last and so is preferred over the original stable one. Leaving us with the expected value, that of the filter.

To illustrate the edge case where this falls over, imagine the following:

1. Core block uses stable support key
2. Plugin applies a filter using the experimental key
3. Theme applies a second filter using the original stable key

Here, the block starts with the stable key and the experimental key is added later by the plugin filter. Using the insertion order to determine/guess which value was updated last fails here as it would indicate the experimental key should be used. The expected outcome would be for the theme's filtered value to be used.

Apologies I wasn't clearer before when referring to the insertion order. Hopefully, I've done a better job communicating this time around.

**Comment by youknowriad** [comment]
I see. Thanks for detailing. I think it's your call if you want to keep the "insertion order" behavior or not (depending on how complex it is). Personally, I'm ok optimizing for the common case which sounds like:

 - plugin disabling stuff using `__experimental`
 - core blocks move toe `stable`

**Comment by aaronrobertshaw** [comment]
@youknowriad I've been attempting to implement a selectors-based approach to stabilizing block supports and it is proving more difficult than hoped for. I'd like to propose a few options to move forward below and would love to get your thoughts.

**Possible options:**

1. Keep the existing approach migrating experimental properties, including default controls, at block type registration
2. Tweak the existing approach but skip stabilizing default controls until they can be a top-level property
3. Revert all the block support stabilization that has already been merged into Gutenberg

**Keeping the existing migrate-on-registration approach**

Pros:

- It's already in place in Gutenberg and seems to work fine
- Keeping `defaultControls` under `supports` means that stabilization can happen straight away rather than wait on core updates for multiple releases
- Most common uses of block support keys e.g. block type support definition, block type metadata filters etc. will continue to work
- Only requires changing experimental keys to stable ones in our codebase instead of changing the method supports are queried (from direct checks to using a selector or util function).

Cons:

- `defaultControls` would still live under `supports`
- Requires any 3rd parties manually checking for formerly experimental support keys to update to the stable key. 
    - This is only an issue for 3rd party plugins that query supports directly on the block type object and try to do something based on that. I struggled to find any use cases matching this when searching WP Directory.
    - We could offset this somewhat by still updating selectors but that comes with its own issues I'll touch on later. I'm not convinced we have to optimise for the above use case
- With the lack of time before 6.8 we probably don't have enough for adequate communication and warning before requiring third party updates.

**Existing migrate-on-registration approach but skip stabilizing defaultControls**

Pros:

- It's pretty much already in place only requiring a small tweak to remove the default controls stabilization
- Doesn't lock `defaultControls` into living under `supports`
- This draft PR already serves as a guide on how we can stabilize `defaultControls` to a top-level property once core supports it

Cons:

- Mostly the same cons as the previous option in that it requires 3rd party updates and communication time


**Revert existing block support stabilization to buy time to find cleaner solution**

Pros:

- We aren't committed yet to any particular direction
- Provides plenty of time to communicate any potential required updates by 3rd parties
- We can focus on more impactful updates for end users in 6.8
- Allows the opportunity to update core now to add `defaultControls` as a top-level block type property which would allow Gutenberg to stabilize default controls in 7.0 and maintain backwards compatibility for the previous two WP versions.

Cons:

- The current approach to stabilization has already been merged and released in Gutenberg
- It's possible some early adopters could have begun updating their code to use stable support keys
- Some communication has recently gone out to the community advertising block support stabilization e.g. the December [What's New for Developers](https://developer.wordpress.org/news/2024/12/whats-new-for-developers-december-2024/#stabilizing-experimental-supports-keys) blog post.
    - This isn't to say we can't backtrack. Just that we have more communication to do to avoid confusion


**Selectors-based Approach**

Here's a bit of a TL;DR on the issues encountered with updating selectors to handle stabilization.

- This approach involves more updates than the migration-on-registration approach, not less
- Trying to retrieve supports from multiple locations looks like it could introduce some performance concerns
- Attempting to support values being in two places at once doesn't really fit with things like our style properties config, global style generation etc that in some parts is still manually looking up values on a supplied block type
- There are as many, or more, places in core to update to use util functions like `block_has_support` than there are uses of experimental keys to simply switch
- It strikes me as easier to convince 3rd parties to switch an experimental key to stable if they are manually querying a block type supports object than getting them to switch to use `getBlockSupport`, `block_has_support` etc.
- Updating multiple selectors e.g. `getBlockSupport`, `getSupportedStyles`, `block_has_support` etc requires a little more untangling if we ever want to be able to remove this code in future. The existing approach only has a 6.8 compat filter, a function within `processBlockType`, and a stabilization function within the block type class in core. 

**Summary**

On their own, most of the issues or cons listed above aren't major. With possibly the exception of needing more time to communicate required changes for 3rd parties. 

I'm genuinely at a bit of a loss as to how to stabilize these supports in a cleaner fashion than we already have in Gutenberg or in any way that won't require 3rd parties to tweak something.

I'll be AFK for the next couple of weeks, into the new year, so the revert option is looking like the best course of action to me right now. I'd be more than happy to hear alternative opinions though!

Given that, I can get a PR together to revert the stabilization of supports already merged in Gutenberg. If we decide to pull the trigger on that, perhaps @ramonjd might be able to hit merge on the revert for me while I'm away 🙏

**Comment by ramonjd** [comment]
Huge kudos to all the work so far @aaronrobertshaw 

And it's great to have our options laid out like this, I appreciate the diligence and clear communication.

> migrating experimental properties, including default controls, at block type registration

Just wanted to note that WordPress has previous form with regards to this approach. See the [wp_migrate_old_typography_shape](https://github.com/WordPress/wordpress-develop/blob/1d49212bb97622e0773652cf1972b86fd08fca22/src/wp-includes/blocks.php#L2360) callback to the `block_type_metadata` hook. It's applied in [register_block_type_from_metadata](https://github.com/WordPress/wordpress-develop/blob/1d49212bb97622e0773652cf1972b86fd08fca22/src/wp-includes/blocks.php#L414).

> defaultControls would still live under supports

What is the basis for this requirement? It seems separate to me. We're talking about stabilization, not API enhancements.

Blocks could happily live without it in the meantime. 

I understand it might make sense to bundle it in with stabilization changes, if it creates too much tech debt later to try to patch in (?).

In that case, it's a solid case for reverting this and doing things more methodically and safely. 

> Revert existing block support stabilization to buy time to find cleaner solution
> With the lack of time before 6.8 we probably don't have enough for adequate communication and warning before requiring third party updates.

I think we should feel comfortable in saying we need more time. 

Nothing is broken, except maybe a few promises made by other folks, no API will change. 

Gutenberg, as every knew from day one, is an experimental plugin, and so long as it's not in Core I think we have licence to revert things that don't make sense, or that jeopardize backwards compatibility.

**Comment by aaronrobertshaw** [comment]
A revert PR for the original approach is up in https://github.com/WordPress/gutenberg/pull/68163.

**Comment by aaronrobertshaw** [comment]
As https://github.com/WordPress/gutenberg/pull/68163 has been merged, I'll close this for the time being.
